### ISSUE TITLE
Proposal: EM_JS - a new way to write JS in C

### ISSUE THREAD

**Issue by jgravelle-google** [emscripten-core/emscripten#6147]
tl;dr: Instead of
```c++
int main() {
  EM_ASM({Module.print("In first function");});
  EM_ASM({Module.print("In second function");});
  EM_ASM({Module.print("Given int: " + $0);}, 101);
  int x = EM_ASM({return 14;});
  EM_ASM({Module.print("Given int: " + $0);}, x);
  return 0;
}
```

we should do something more like:
```c++
EM_JS(void test_void_no_args(), {
  Module.print("In first function");
});
EM_JS(void test_void_void_arg(void), {
  Module.print("In second function");
});
EM_JS(void print_int(int x), {
  Module.print("Given int: " + x);
});
EM_JS(int test_int_no_args(), {
  return 14;
});

int main() {
  test_void_no_args();
  test_void_void_args();
  print_int(101);
  int x = test_int_no_args();
  print_int(x);
  return 0;
}
```

## Motivation
The ways specified in the [Interacting with code](https://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html) documentation for calling JS functions from C are emscripten_run_script (which uses eval and is less fully-featured), EM_ASM, or [adding a JS library](https://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#implement-a-c-api-in-javascript).
For small self-contained example programs, EM_ASM is the most attractive option, it's flexible, it's all in one file, it doesn't require changing the build command. It'd be nice if we could use the library-function approach with a similar level of ease as EM_ASM.

## Enter EM_JS
The idea is essentially to automatically generate a named JS library function for the user from C. Similar to EM_ASM, with a few advantages:
1) User provides a C-side function signature, removing ambiguity at the callsite
2) User provides argument names, no longer needing to count $0 $1 $2 etc, which should be more ergonomic
3) The syntax more closely matches what the implementation actually does - declares a function with a C signature and JS implementation
4) Generated JS functions can be direct imports, instead of a call through an index into a JS array of functions. This is more slightly performant, though the wasm->JS call overhead is still present. Code size should be slightly smaller because we can avoid the __emscripten_asm_const_[sig] thunks

## Example
Starting with the user-provided C source
```c++
EM_JS(float foo(int x, float y), {
  var product = x * y;
  Module.print('Given (' + x + ', ' + y + '), product is: ' + product);
  return product;
});
```
turn that (via the preprocessor) into
```c++
float foo(int x, float y);
```
and in wasm,
```wasm
(import "env" "foo" (function $foo (param i32 f32) (result f32)))
;; ...
(call $foo
  (i32.const 5)
  (f32.const 9)
)
```
with JS glue (in the import env)
```js
"foo": function(x, y) {
  Module.print('Given (' + x + ', ' + y + '), product is: ' + product);
  return product;
},
```

## Implementation
First idea that comes to mind:
1) Use a C macro to quote the entire function declaration and body into a string, put the string in the wasm data section [via some tbd mechanism, this is trickier to do from outside a function than I'd thought]
2) Binaryen fishes out the pattern in the same way we handle EM_ASM today, adds it to its own section in metadata
3) Emscripten reads the function bodies and generates the imports

## EM_ASM deprecation?
If we do this, we'll have two features in the platform that essentially do the same thing - embed JS in C, and automatically generate function import glue. I believe EM_JS provides a nicer EM_ASM-like experience for users, so we should encourage its use (And if it isn't nicer than EM_ASM, we probably shouldn't implement it at all). We shouldn't remove EM_ASM, but we should move it to a footnote in the documentation, and try to remove it from future example code.

**Comment by juj** [comment]
Like the new syntax! All looks good, except for the deprecation part - let's not rush into downplaying it, there should not be much reason not to, even if there are inefficiencies related. These two syntaxes could be seen as regular (`EM_JS`) vs lambda (`EM_ASM`) functions, and deprecating `EM_ASM` would be mentally akin to deprecating lambda functions that C++11 added. A common usage is to do something like

```c
int foo()
{
   double devicePixelRatio = EM_ASM_DOUBLE(return window.devicePixelRatio);
   int width = EM_ASM_INT(return screen.width);
   int height = EM_ASM_INT(return screen.height);
   ...
}
```

which is super lean for developers to grok.

Would LLVM be able to treat these identical to regular C/C++ functions with respect to linkage and declaration/definitions? E.g.

```c
foo.c

EM_JS(float foo(int x, float y), {
  var product = x * y;
  Module.print('Given (' + x + ', ' + y + '), product is: ' + product);
  return product;
});
```

```c
foo.h

float foo(int x, float y);
```

```c
main.c

#include "foo.h"

int main()
{
   float y = foo(1, 4.f); // User doesn't need to care or see if foo was implemented in JS or C/C++?
}
```

Another random thought is - would it be possible to teach LLVM with an attribute, like

```
float __attribute__((language=javascript)) foo(int x, float y)
{
  var product = x * y;
  Module.print('Given (' + x + ', ' + y + '), product is: ' + product);
  return product;
}
```

Not too important, but that would look pretty cool, especially if the linkage rules can be made identical to C/C++ functions?

With respect to multithreading, there is the question of how the thread context is specified. Either the definition would specify, call site would specify, both would be allowed to specify, or neither. I'd be leaning towards allowing both function definition and call site to be able to specify the thread context, to ease the flexibility in defining proxying models as much as possible, but that can be discussed more further to not complicate here.

**Comment by kripken** [comment]
I like this a lot!

Could c++11 lambdas allow writing EM_JSes inline?

**Comment by tbfleming** [comment]
Here's a way to get lambda-like behavior, but without captured variables. If there's interest, there may be a way to add captured variable support.

```
template<typename R, typename... T>
R js_function(const char* sig, const char* code, T...);

template<typename T>
struct js_lambda;

template<typename R, typename... T>
struct js_lambda<R(T...)> {
    const char* const sig;
    const char* const code;
    R operator()(T... args) const {
        return js_function<R, T...>(sig, code, args...);
    }
};

#define JS_LAMBDA(sig, code) (js_lambda<sig>{#sig, #code})
```

binaryen would examine calls to mangled js_function. The first two args are the function signature and the function body. It would replace the calls to js_function with calls to a function it generates.

Example usage:

```
int main() {
    // lambda as object
    auto lambda = JS_LAMBDA(double(int i, float j, const char* p), {
        console.log(i, j, p);
        return 7.2;
    });
    printf("%f\n", lambda(5, 2.1, "test 1"));

    // immediately invoked
    double d = JS_LAMBDA(double(int i, float j, const char* p), {
        console.log(i, j, p);
        return 4.9;
    })(2, 3.7, "test 2");
    printf("%f\n", d);
}
```

This generates this wasm:
```
...
      (call $_Z11js_functionIdJifPKcEET_S1_S1_DpT0_
        (i32.const 0)
        (i32.const 38)
        (i32.const 5)
        (f32.const 0x1.0cccccp+1 (;=2.1;))
        (i32.const 80)))
...
      (call $_Z11js_functionIdJifPKcEET_S1_S1_DpT0_
        (i32.const 0)
        (i32.const 87)
        (i32.const 2)
        (f32.const 0x1.d9999ap+1 (;=3.7;))
        (i32.const 125)))
...
  (data (i32.const 0) "double(int i, float j, const char* p)\00")
  (data (i32.const 38) "{ console.log(i, j, p); return 7.2; }\00")
...
  (data (i32.const 87) "{ console.log(i, j, p); return 4.9; }\00")
```

**Comment by jgravelle-google** [comment]
@juj 
> These two syntaxes could be seen as regular (EM_JS) vs lambda (EM_ASM) functions

I hadn't thought of EM_ASM being like a lambda before. That makes sense.
For deprecation, I was mostly thinking "prefer recommending EM_JS instead of EM_ASM going forward, change documentation to reflect that." I don't think we'll want to drop tools support for EM_ASM.
My goal with EM_JS is to subsume 90+% of the use cases for EM_ASM by being the more convenient option, e.g. if your code has lots of `EM_ASM(return screen.height)` calls, it might be nicer to have a `screenHeight()` helper function, and if your function body is just a call to EM_ASM, EM_JS seems more natural.

> even if there are inefficiencies related

The tone I tried to take with respect to the efficiencies was that this will be slightly more efficient, but not significantly enough to be a deciding factor. The main selling point to me is ergonomics, but we're not *spending* performance here to do so.

> Would LLVM be able to treat these identical to regular C/C++ functions with respect to linkage and declaration/definitions?

That's exactly the plan. That's the part that makes it more natural for the tools to consume, and as far as user code is concerned it's just another C function.

> would it be possible to teach LLVM with an attribute

I had that thought too. I'm highly doubtful that it would be, because we'd need to teach the Clang frontend, and in particular we'd need support in the parser, if not the lexer, and I'm not sure if attributes are even read by that time.
Additionally one of my secondary goals with EM_JS is to make something that's easier to implement in the toolchain without needing to add Emscripten-specific logic to LLVM.

@kripken 
>Could c++11 lambdas allow writing EM_JSes inline?

Highly unlikely. The idea is still to quote the JS function body as a string, stash it in a wasm data section, then postprocess it in Binaryen. We could manually generate lambda-like objects maybe, but I'm not sure how.

@tbfleming 
... Ok maybe like that. Nice! Very clever. I'd love to take it one step further and have something like
```c++
template<typename R, const char Code[], typename... T>
R js_function(const char* sig, T...);
```
which would cause the template system to generate a separate function for each code string, which would let us use the direct function calling mechanism of EM_JS, as opposed to the indirect function table mechanism of EM_ASM.
... Though it's highly likely there may be no getting around that, because using strings as template parameters doesn't quite work the way I'd hope - I tried something similar with EM_ASM and came up empty.

**Comment by tbfleming** [comment]
Each time binaryen encounters a js_function invocation, it could follow the first 2 arguments (hopefully llvm will always make them constants), look up {sig, code}, and generate a function if it doesn't already exist. Once it has that function it could replace the call to js_function with a call to the generated function. This would produce direct calls; no indirect table needed.

**Comment by tbfleming** [comment]
By generate, I mean generate a JS function and a matching import.

**Comment by juj** [comment]
> For deprecation, I was mostly thinking "prefer recommending EM_JS instead of EM_ASM going forward, change documentation to reflect that." I don't think we'll want to drop tools support for EM_ASM.
> My goal with EM_JS is to subsume 90+% of the use cases for EM_ASM by being the more convenient option

Agreed on all points.

> if your code has lots of EM_ASM(return screen.height) calls, it might be nicer to have a screenHeight() helper function, and if your function body is just a call to EM_ASM, EM_JS seems more natural.

This also makes sense. I've seen a number of cases where developers have practically the identical `EM_ASM` block several times, but in a tiny bit different form, which would result in duplicate identical forms of those `EM_ASM` function blocks generated. This is why the --js-library approach has been superior for bigger cases, much what this is achieving as well.

> The tone I tried to take with respect to the efficiencies was that this will be slightly more efficient, but not significantly enough to be a deciding factor. The main selling point to me is ergonomics, but we're not spending performance here to do so.

Gotcha, +1.

I really like this, especially that with this approach, there is an opportunity to strongly define the signatures of the input arguments on int vs double vs int64 and so on, which is a synergy that proxying to threads can leverage, as well as perhaps if in the future we wanted to directly hook in high level marshalling of e.g. strings or similar.

**Comment by jgravelle-google** [comment]
This was implemented in #6220, closing
